refactor(compiler): add support for host directives

crisbeto opened the pull request:
This is the compile-time implementation of the `hostDirectives` feature plus a little bit of runtime code to illustrate how the newly-generated code will plug into the runtime. It works by creating a call to the new `ɵɵHostDirectivesFeature` feature whenever a directive has a `hostDirectives` field. Afterwards `ɵɵHostDirectivesFeature` will patch a new function onto the directive definition that will be invoked during directive matching.

For example, if we take the following definition:

```ts
@Directive({
  hostDirectives: [HostA, {directive: HostB, inputs: ['input: alias']}]
})
class MyDir {}
```

Will compile to:

```js
MyDir.ɵdir = ɵɵdefineComponent({
  features: [ɵɵHostDirectivesFeature([HostA, {
    directive: HostB,
    inputs: ['input', 'alias']
  }])]
});
```

The template type checking is implemented during directive matching by adding the host directives applied on the host to the array of matched directives whenever the host is matched in a template.

Relates to #8785.

eneajaho commented:
```ts
const TOKEN = new InjectionToken("Test", {
  providedIn: "root",
  factory: () => 'From token'
})

@Directive()
class MyDir {
  constructor(@Inject(TOKEN) public token: string) {
    console.log(token === 'From component!'); // will this be true ?
  }
}

@Component({
  providers: [{ provide: TOKEN, useValue: 'From component' }],
  hostDirectives: [MyDir]
})
export class MyCmp {}

```

HostDirectives will live in the same node injector as the component, right? 
So, `console.log(token === 'From component!');` will log true, right?

crisbeto commented:
Yes, you can imagine the example above as representing `<my-cmp my-dir></my-cmp>`.

nartc commented:
Awesome!

How does this work with Structural Directives? (do they work? I'm not sure since Structural Directives aren't mentioned in the PR)

```ts
@Component({
  selector: 'some'
  hostDirectives: [
     { directive: NgIf, inputs: ['ngIf: show'] }
  ]
})
export class SomeComponent {
   @Input() show = true;
}
```
```html
<some [show]="shouldShow"></some>
```

- Does this even work? since `*ngIf` is syntactic sugar for `<ng-template ngIf>` so it might not work?
- Does `<some>` get taken out of the DOM like `<some *ngIf=""></some>`

crisbeto commented:
You'll likely get a DI error if you tried to set `NgIf` as a host directive on a non-template element, because it won't be able to inject `TemplateRef`.

nartc commented:
@crisbeto thank you for the response. I figured it'd be the case. In that case, should we update the PR description to call it out for Structural Directives? I guess it makes sense that this feature only targets Attribute Directive. Structural should be handled on the consuming (component's users) side of things rather than authoring (component's author).

Another question: is it possible to have TypeScript generics supported?

```ts
@Directive()
export abstract class NgtInstance<TObject> {
   @Output() ready = new EventEmitter<TObject>();
}

@Component()
export class NgtMesh extends NgtInstance<THREE.Mesh> {}
```

Here's a sample inheritance example. In reality, the inheritance level might be more than 1 level deep. You can see here that I can have generics with inheritance, and `(ready)` is strongly typed to `THREE.Mesh`

Let's try with Directive Composition

```ts
@Directive({
   selector: '[ngt-instance]'
   standalone: true
})
export class NgtInstance<TObject> {
   @Output() ready = new EventEmitter<TObject>();
}

@Component({
   hostDirectives: [
      // not sure how to provide THREE.Mesh here for (ready) to be strongly typed
      // maybe something like NgtInstance as NgtInstance<THREE.Mesh> and the compiler can figure it out?
      { directive: NgtInstance, outputs: ['ready'] }
   ]
})
export class NgtMesh {}
```

crisbeto commented:
Can we move all of the discussions to https://github.com/angular/angular/issues/8785? Let's keep the PR only for actual reviews.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -614,3 +626,88 @@ function evaluateHostExpressionBindings(
+ */
+function extractHostDirectives(
+    hostDirectivesExpression: ts.Expression, evaluator: PartialEvaluator, context: ts.SourceFile,
+    reflector: ReflectionHost) {
```
Nit: explicit return type

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -614,3 +626,88 @@ function evaluateHostExpressionBindings(
+    let hostReference = value instanceof Map ? value.get('directive') : value;
+    let forwardRefNode: ts.Node|null = null;
+
+    if (hostReference instanceof DynamicValue && ts.isCallExpression(hostReference.node)) {
```
This is definitely not how we should look for usages of `forwardRef` (for example, the user could `import * as core from '@angular/core'` and call `forwardRef` as `core.forwardRef()`).

The theory of the `PartialEvaluator` is that we should just be able to pass `forwardRefResolver` to the top-level evaluation above, and it should just work - we'll get a `Reference` in place of the `forwardRef` call.

crisbeto replied:
I did feel iffy when I was writing this, but the problem I was trying to solve is that I wanted to preserve the `forwardRef` expression in the generated code (e.g. `HostDirectivesFeature([forwardRef(() => HostDir)])`), whereas passing in the `forwardRefResolver` would resolve to the class declaration.

alxhub replied:
We typically don't leave `forwardRef` in generated expressions like this, but instead generate a closure for the whole structure that might contain forward references:
```typescript
HostDirectivesFeature(() => [HostDir])
```
This is a bit more compact in terms of generated code, and easier to resolve at runtime than wrapping reads in `resolveForwardRef`.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -614,3 +626,88 @@ function evaluateHostExpressionBindings(
+ */
+function parseHostDirectivesMapping(
+    field: 'inputs'|'outputs', resolvedValue: ResolvedValue, classReference: ClassDeclaration,
+    sourceExpression: ts.Expression) {
```
Nit: explicit return type

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -614,3 +626,88 @@ function evaluateHostExpressionBindings(
+      inputs: ClassPropertyMapping|null,
+      outputs: ClassPropertyMapping|null
+    },
+    context: ts.SourceFile, refEmitter: ReferenceEmitter) {
```
Same here

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -496,7 +500,7 @@ export class ComponentDecoratorHandler implements
     const scope = this.scopeReader.getScopeForComponent(node);
     const selector = analysis.meta.selector;
-    const matcher = new SelectorMatcher<DirectiveMeta>();
+    const matcher = new SelectorMatcher<DirectiveMeta[]>();
```
I think we need to distinguish between matched directives and host directives here. The template type-checker will need to care about the difference, because only a subset of host directive inputs/outputs are legal to bind, and may be renamed. I don't think this is captured in the template type-checking tests at the moment. If you add such a test (and we should have one), I suspect it would fail.

crisbeto replied:
The input/output filtering happens in `HostDirectivesResolver.filterMappings`. There are tests for it in `template_typecheck_spec.ts`: https://github.com/angular/angular/pull/46868/files#diff-4338bcab035efc17fedac8da7ef2cdfaa3f171d673ed2cd22cbe42641c998fb1R2798

alxhub reviewed (changes requested):
First round of review! Overall this is in pretty good shape.

A few meta-comments about the PR as a whole:
* Ideally, we would split this up into multiple commits.

For example, one commit could deal with adding the extra metadata to .d.ts files (starting as `never` always), another could do the parsing, another the template type-checking, etc.

* This should not be a `feat` commit if the feature is not actually implemented.

Instead, make it a `refactor` commit. We don't want any entry in the changelog until the feature is usable to users. For a big implementation like this, we can have multiple PRs adding implementation details that aren't exposed, and then a final `feat` commit which only makes the API public.

* On that note, we shouldn't also add `hostDirectives` to the public `@Directive` metadata just  yet.

That is, we shouldn't change the public API and allow users to access the feature until it's fully implemented and working.

crisbeto commented:
Regarding splitting it up into multiple commits: I agree that this is a huge change, but I ended up going with the single commit, because the majority of it are the actual tests, not the compiler code, and in order to make some tests work, most of the compiler changes would have to be in place.

Regarding the commit message: I can target this to the `main` branch so it won't show up in the changelog until we release it. Alternatively I can target it to a feature branch instead.

Regarding exposing the new API in the `@Directive` metadata: I exposed it, because some of the tests are basically testing the entire pipeline end-to-end and I wanted to avoid compilation errors. The code won't be released for a while and even if it was passed in, the runtime isn't in place so nothing will happen.

crisbeto committed:
refactor(compiler): add support for host directives

This is the compile-time implementation of the `hostDirectives` feature plus a little bit of runtime code to illustrate how the newly-generated code will plug into the runtime. It works by creating a call to the new `ɵɵHostDirectivesFeature` feature whenever a directive has a `hostDirectives` field. Afterwards `ɵɵHostDirectivesFeature` will patch a new function onto the directive definition that will be invoked during directive matching.

For example, if we take the following definition:

```ts
@Directive({
  hostDirectives: [HostA, {directive: HostB, inputs: ['input: alias']}]
})
class MyDir {}
```

Will compile to:

```js
MyDir.ɵdir = ɵɵdefineComponent({
  features: [ɵɵHostDirectivesFeature([HostA, {
    directive: HostB,
    inputs: {
      input: "alias"
    }
  }])]
});
```

The template type checking is implemented during directive matching by adding the host directives applied on the host to the array of matched directives whenever the host is matched in a template.

Relates to #8785.

crisbeto committed:
fixup! refactor(compiler): add support for host directives

crisbeto commented:
@alxhub I've addressed the feedback above and the things we talked about in person. All the changes are in the second fixup commit. Here's a high level overview of what has changed:
* The commit message was switched to `refactor` so it doesn't end up in the changelog until the feature has been implemented completely.
* Instead of preserving `forwardRef` calls, the compiler will generate a function when there are forward references.
* `hostDirectives` has been removed from the public API of `@Directive` and `@Component` until we're done implementing the feature.
* I've added some logic to filter out the host directives from the component template dependencies, as we discussed.
* The compiler now produces diagnostics for problems in the host directives definition, instead of throwing errors. Also the validation happens in the `validate` method of `HostDirectivesResolver` instead of in `resolve`.
* I had to expand the type on a couple of utilities that deal with diagnostics from `ts.Expression` to `ts.Node` so that I could produce diagnostics with the correct context for host directives. As far as I can tell, we weren't using anything specific to `Expression` anyway.
* The comments around return types and moving literals into dedicated types have been resolved.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/host_directives_resolver.ts:
```diff
@@ -0,0 +1,116 @@
+  private walkHostDirectives(
+      directives: NonNullable<DirectiveMeta['hostDirectives']>, results: T[]): ReadonlyArray<T> {
+    for (const current of directives) {
+      const hostMeta = this.metaReader.getDirectiveMetadata(current.directive);
```
This will only produce shallow metadata - if `current.directive` is a subclass of another directive, for example, the superclass' inputs/outputs won't be included in the `hostMeta.inputs` that get returned here (or filtered properly).

That is, the resulting metadata here won't be suitable for template type-checking, as we won't have a correct set of inputs.

Can you add a test to confirm this? I suspect we'll have to adjust the interface here to accept a `(ref: Reference<ClassDeclaration>) => DirectiveMeta` resolver function, so we can use the correct template type-checking metadata resolver from `TypeCheckScopeRegistry` when required.

crisbeto replied:
You're totally right, good catch! I did end up doing it slightly differently by using `flattenInheritedDirectiveMetadata` inside the `HostDirectivesResolver` so that consumers of the class don't have to know about what it is using internally to resolve the host directives.

crisbeto committed:
fixup! refactor(compiler): add support for host directives

crisbeto commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -118,6 +118,11 @@ export interface R3DeclareDirectiveMetadata extends R3PartialDeclaration {
+  /**
+   * Additional directives applied to the directive host.
+   */
+  hostDirectives?: (o.Expression|R3DeclareHostDirectiveMetadata)[]|o.FunctionExpr;
```
I think that this would actually be harder to maintain, because we reuse the same `createHostDirectivesArg` function both for partial and full compilation. For reference, the compiled output looks like this:

```
// Full, no forward refs
features: [HostDirectivesFeature([DirA, {directive: DirB, inputs: ['foo']}])]

// Full with forward refs
features: [HostDirectivesFeature(() => [DirA, {directive: DirB, inputs: ['foo']}])]

// Partial, no forward refs
ngDeclareDirective({
  hostDirectives: [DirA, {directive: DirB, inputs: ['foo']}]
});

// Partial with forward refs
ngDeclareDirective({
  hostDirectives: () => [DirA, {directive: DirB, inputs: ['foo']}]
})
```

alxhub replied:
Yes, I understand that we wouldn't be able to reuse the same code to create the argument. However, the goal of the partial compilation format is not to be as close to the final generated code as possible, but to encode the semantics of the user's intentions as cleanly as possible. This allows future compilers to know what code to generate, even if the final compiled form of that code changes.

crisbeto replied:
I've pushed a change to always output `R3DeclareHostDirectiveMetadata` and to generate a `forwardRef` for forward references.

crisbeto commented:
@alxhub I've addressed the latest set of feedback. All of the new code is in the latest `fixup` commit.

crisbeto committed:
fixup! refactor(compiler): add support for host directives

crisbeto commented:
I've pushed an additional commit to fix the issue where host directives weren't accounting for aliases. E.g. if we take the following example:

```ts
@Directive({
  standalone: true
})
class HostDir {
  @Input('alias') input: any;
}

@Directive({
  hostDirectives: [{directive: HostDir, inputs: ['alias: customAlias']}]
})
class Host {}
```

Before the latest commit, `alias: customAlias` wouldn't have worked, because we were only looking at the private name. In the process of fixing it, I realized that the previous compiler output was going to break with property renaming. The problem was that the call to the feature would look something like `HostDirectivesFeature([{directive: Foo, inputs: {publicName: 'hostDirectiveAlias'}}])`. The keys in the object literal in `inputs` always have to refer to the public name which would've been minified by something like Closure Compiler. In the latest commit the compiler outputs the following instead: `HostDirectivesFeature([{directive: Foo, inputs: ['publicName', 'publicAlias']}])` which then gets converted into an object literal at runtime.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/host_directives_resolver.ts:
```diff
@@ -0,0 +1,128 @@
+      directives: NonNullable<DirectiveMeta['hostDirectives']>,
+      results: DirectiveMeta[]): ReadonlyArray<DirectiveMeta> {
+    for (const current of directives) {
+      const hostMeta = flattenInheritedDirectiveMetadata(this.metaReader, current.directive);
```
I see that this is now flattening inherited metadata, which addresses the problem with template type-checking not considering metadata from inheritance. However, it's now _also_ flattening metadata for the main directive metadata, which may cause information from inherited metadata to be emitted into generated code.

Here's why that's a problem. Let's say `DirectiveB` inherits from `DirectiveA`, which is published separately. If we include metadata from `DirectiveA` in our compilation of `DirectiveB`, and then a new version of `DirectiveA` is published and installed by the application which invalidates that metadata, then the shipped version of `DirectiveB` might be incorrect.

If we do make use of inherited data, we need to be very careful that it's public API data only - something that would be a breaking change for `DirectiveA` to update. And we should be clear that we're doing this in our design.

For most of the code we generate, we instead only consider information we learned from the local class declaration. The runtime is responsible for synthesizing information from a superclass if there is one. Host directives should behave this way as well: we should only use information from superclasses at runtime (and for template type-checking).

crisbeto committed:
fixup! refactor(compiler): add support for host directives

bryanrideshark commented:
This is the most exciting improvement Angular has had since Ivy!

I'm itching to know more and many others are as well I'm sure. Is there a design document we can look at to familiarize ourselves with the particulars?

bryanrideshark commented:
Can we do the following?

````

@Directive({
  standalone: true
})
class HostDir {
  @Input('alias') input: any;
  
  configureProgrammatically() {
     // Maybe some complex things that really would be best done via TypeScript
  }
}

@Directive({
  hostDirectives: [{directive: HostDir, inputs: ['alias: customAlias']}]
})
class Host {
    constructor(public host: HostDir) {}
    
    ngOnInit() { 
       this.host.configureProgrammatically();
    }
}

````

alxhub commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts:
```diff
@@ -154,6 +157,72 @@ function toQueryMetadata<TExpression>(obj: AstObject<R3DeclareQueryMetadata, TEx
+ */
+function toHostDirectivesMetadata<TExpression>(
+    hostDirectives: AstValue<
+        (o.Expression | R3DeclareHostDirectiveMetadata)[]|o.FunctionExpr|undefined, TExpression>):
```
Partially compiled host directive declarations should _always_ be an array of `R3DeclareHostDirectiveMetadata`, so we should be able to drop the `o.Expression` and `o.FunctionExpr` form:
```suggestion
        R3DeclareHostDirectiveMetadata[]|undefined, TExpression>):
```

The forward reference should be tracked individually (the `directive` field of each one might be a function).

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/common/src/diagnostics.ts:
```diff
@@ -151,6 +151,39 @@ export function getDirectiveDiagnostics(
+      diagnostics.push(makeDiagnostic(
+          ErrorCode.UNRESOLVED_HOST_DIRECTIVE,
+          current.directive.getOriginForDiagnostics(current.origin),
+          `Could not resolve host directive metadata of ${current.directive.debugName}`));
```
This reads more like an internal error. Instead we should say something like `${current.directive.debugName} must be a standalone directive to be used as a host directive`

alxhub commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -71,6 +71,21 @@ export enum ErrorCode {
+  /**
+   * Raised when the compiler wasn't able to resolve the metadata of a host directive.
+   */
+  UNRESOLVED_HOST_DIRECTIVE = 2013,
```
```suggestion
  HOST_DIRECTIVE_INVALID = 2013,
```

alxhub commented on packages/compiler-cli/src/ngtsc/imports/src/references.ts:
```diff
@@ -153,9 +153,8 @@ export class Reference<T extends ts.Node = ts.Node> {
-  getOriginForDiagnostics(container: ts.Expression, fallback: ts.Expression = container):
-      ts.Expression {
-    const id = this.getIdentityInExpression(container);
+  getOriginForDiagnostics(container: ts.Node, fallback: ts.Node = container): ts.Node {
```
These changes shouldn't be necessary - `getOriginForDiagnostics` should be able to only accept `ts.Expression`s as that's the only context in which we should show diagnostics.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/api.ts:
```diff
@@ -161,6 +175,27 @@ export interface DirectiveMeta extends T2DirectiveMeta, DirectiveTypeCheckMeta {
+  directive: Reference<ClassDeclaration>;
+
+  /** Node in which the host directive was declared. Used for diagnostics. */
+  origin: ts.Node;
```
I think we can drop this field. See my comment on `getHostDirectiveDiagnostics`.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/common/src/diagnostics.ts:
```diff
@@ -151,6 +151,39 @@ export function getDirectiveDiagnostics(
   return diagnostics;
 }
 
+export function getHostDirectiveDiagnostics(
```
```suggestion
export function validateHostDirectives(
```

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/common/src/diagnostics.ts:
```diff
@@ -151,6 +151,39 @@ export function getDirectiveDiagnostics(
+    if (hostMeta === null) {
+      diagnostics.push(makeDiagnostic(
+          ErrorCode.UNRESOLVED_HOST_DIRECTIVE,
+          current.directive.getOriginForDiagnostics(current.origin),
```
This function is called by the component/directive handlers to validate that the host directives applied by the user are correct. All of the host directives here will come from a single component/directive `hostDirective` expression, which we should pass in as a parameter (in addition to `hostDirectives`) and use as the origin expression for all of the diagnostics.

Given that, I don't think we need to track a separate `origin` within `HostDirectiveMeta`.

This resolves the weirdness around needing `origin` even for host directives extracted from .d.ts metadata, which doesn't have `ts.Expression`s. There's no way we'll ever end up showing diagnostics for such directives - they've already been validated by the upstream compilation.

We do something similar with other fields - `NgModuleAnalysis` for example includes a number of `ts.Expression` fields for `rawImports`, `rawExports`, etc. which are used for such diagnostics.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/test/util.ts:
```diff
@@ -44,19 +44,19 @@ export function getBoundTemplate(
+                             exportAs: null,
+                             isStructural: false,
+                             animationTriggerNames: null,
+                           }]);
```
🤦 this formatter...

alxhub commented on packages/compiler-cli/src/ngtsc/scope/src/util.ts:
```diff
@@ -15,7 +15,7 @@ import {ClassDeclaration} from '../../reflection';
 
 export function getDiagnosticNode(
-    ref: Reference<ClassDeclaration>, rawExpr: ts.Expression|null): ts.Expression {
+    ref: Reference<ClassDeclaration>, rawExpr: ts.Expression|null): ts.Node {
```
Also seems unnecessary.

alxhub commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -812,3 +819,78 @@ function compileStyles(styles: string[], selector: string, hostSelector: string)
   });
 }
+
+function createHostDirectivesType(meta: R3DirectiveMetadata) {
```
Explicit return types on functions here

alxhub commented on packages/core/src/metadata/directives.ts:
```diff
@@ -332,6 +332,20 @@ export interface Directive {
    * @developerPreview
    */
   standalone?: boolean;
+
```
We can omit commented-out code for now. Especially since this dangling `/**` comment doesn't have a property it's attached to.

alxhub commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -416,6 +416,12 @@ export function directiveMetadata(type: Type<any>, metadata: Directive): R3Direc
     viewQueries: extractQueriesMetadata(type, propMetadata, isViewQuery),
     isStandalone: !!metadata.standalone,
+
+    // TODO(crisbeto): enable this when implementing the host directives runtime.
```
Leave out commented-out code for now.

crisbeto committed:
fixup! refactor(compiler): add support for host directives

crisbeto committed:
fixup! refactor(compiler): add support for host directives

crisbeto commented:
The latest round of feedback has been addressed.

alxhub commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -812,3 +819,78 @@ function compileStyles(styles: string[], selector: string, hostSelector: string)
+}
+
+function createHostDirectivesFeatureArg(
+    hostDirectives: NonNullable<R3DirectiveMetadata['hostDirectives']>) {
```
Nit: explicit return type

alxhub commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1311,6 +1311,10 @@ function findDirectiveDefMatches(
         }
+
+        if (def.applyHostDirectives) {
+          def.applyHostDirectives(tView, viewData, tNode, matches);
```
We could use the fancy `def.applyHostDirectives?.(...)` here to save a few bytes :)

alxhub reviewed (approved):
Reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

crisbeto committed:
fixup! refactor(compiler): add support for host directives

jessicajaniuk reviewed (approved):
reviewed-for: public-api

alxhub commented:
This PR was merged into the repository by commit 54ceed53e21a6a899617b626becc51ffb1059178.
